kwave/utils/filters.py

```diff
@@ -233,37 +233,63 @@ def extract_amp_phase(data: np.ndarray, Fs: float, source_freq: float, dim: Tupl
 
 
 def brenner_sharpness(im):
-    ndim = len(np.squeeze(im).shape)
-
-    if ndim == 2:
+    num_dim = im.ndim
+    if num_dim == 2:
+        # compute metric
         bren_x = (im[:-2, :] - im[2:, :]) ** 2
         bren_y = (im[:, :-2] - im[:, 2:]) ** 2
-        s = bren_x.sum() + bren_y.sum()
-    elif ndim == 3:
+        s = np.sum(bren_x) + np.sum(bren_y)
+    elif num_dim == 3:
+        # compute metric
         bren_x = (im[:-2, :, :] - im[2:, :, :]) ** 2
         bren_y = (im[:, :-2, :] - im[:, 2:, :]) ** 2
         bren_z = (im[:, :, :-2] - im[:, :, 2:]) ** 2
-        s = bren_x.sum() + bren_y.sum() + bren_z.sum()
-    else:
-        raise ValueError("Invalid number of dimensions in im")
+        s = np.sum(bren_x) + np.sum(bren_y) + np.sum(bren_z)
     return s
 
 
 def tenenbaum_sharpness(im):
-    ndim = len(np.squeeze(im).shape)
-    if ndim == 2:
-        sobel = scipy.ndimage.sobel(im)
-    elif ndim == 3:
-        sobel = scipy.ndimage.sobel(im)
+    num_dim = im.ndim
+    if num_dim == 2:
+        # define the 2D sobel gradient operator
+        sobel = np.array([[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]])
+
+        # compute metric
+        s = (convolve(sobel, im) ** 2 + convolve(sobel.T, im) ** 2).sum()
+    elif num_dim == 3:
+        # define the 3D sobel gradient operator
+        sobel3D = np.zeros((3, 3, 3))
+        sobel3D[:, :, 0] = np.array([[1, 2, 1], [2, 4, 2], [1, 2, 1]])
+        sobel3D[:, :, 2] = -sobel3D[:, :, 0]
+
+        # compute metric
+        s = (convolve(im, sobel3D) ** 2 + convolve(im, np.transpose(sobel3D, (2, 0, 1))) ** 2 +
+             convolve(im, np.transpose(sobel3D, (1, 2, 0))) ** 2).sum()
+    return s
 
-    else:
-        raise ValueError("Invalid number of dimensions in im")
-    return sobel.sum()
+    # NOTE: Walter thinks this is the proper way to do this, but it doesn't match the MATLAB version
```

Comment:
Looks neat, I vote for implementing Walter's version :)

Reply:
in the future...